timely/src/dataflow/channels/mod.rs

```diff
@@ -36,10 +36,25 @@ impl<T, D> Message<T, D> {
         Message { time, data, from, seq }
     }
 
-    /// Forms a message, and pushes contents at `pusher`.
+    /// Forms a message, and pushes contents at `pusher`. Replaces `buffer` with what the pusher
+    /// leaves in place, or a new `Vec`. Note that the returned vector is always initialized to
+    /// contain [Self::default_length] elements.
     #[inline]
     pub fn push_at<P: Push<Bundle<T, D>>>(buffer: &mut Vec<D>, time: T, pusher: &mut P) {
 
+        Self::push_at_no_allocation(buffer, time, pusher);
+
+        // TODO: Unclear we always want this here.
+        if buffer.capacity() != Self::default_length() {
+            *buffer = Vec::with_capacity(Self::default_length());
+        }
+    }
+
+    /// Forms a message, and pushes contents at `pusher`. Replaces `buffer` with what the pusher
+    /// leaves in place, or a new empty `Vec`.
```

Comment:
I think the docs can be a bit more specific, that the new empty vec occurs only when the returned buffer is larger than the default size.

Reply:
Added a sentence.